libbpf-tools/cpufreq.c

```diff
@@ -0,0 +1,254 @@
+// 10-OCT-2020   Wenbo Zhang   Created this.
+#include <argp.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <linux/perf_event.h>
+#include <asm/unistd.h>
+#include <bpf/libbpf.h>
+#include <bpf/bpf.h>
+#include "cpufreq.h"
+#include "cpufreq.skel.h"
+#include "trace_helpers.h"
+
+static struct env {
+	int duration;
+	int freq;
+	bool verbose;
+} env = {
+	.duration = -1,
+	.freq = 99,
+};
+
+const char *argp_program_version = "cpufreq 0.1";
+const char *argp_program_bug_address = "<bpf@vger.kernel.org>";
+const char argp_program_doc[] =
+"Sampling CPU freq system-wide & by process. Ctrl-C to end.\n"
+"\n"
+"USAGE: cpufreq [--help] [-d DURATION] [freq]\n"
+"\n"
+"EXAMPLES:\n"
+"    cpufreq         # sample CPU freq as 99hz (default)\n"
+"    cpufreq -d 5    # sample for 5 seconds only\n"
+"    cpufreq 199     # sample CPU freq as 199hz\n"
+"\n";
+
+static const struct argp_option opts[] = {
+	{ "duration", 'd', "DURATION", 0, "Duration to trace in seconds" },
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	static int pos_args;
+
+	switch (key) {
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'd':
+		errno = 0;
+		env.duration = strtol(arg, NULL, 10);
+		if (errno || env.duration <= 0) {
+			fprintf(stderr, "Invalid duration: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case ARGP_KEY_ARG:
+		if (pos_args++) {
+			fprintf(stderr,
+				"Unrecognized positional argument: %s\n", arg);
+			argp_usage(state);
+		}
+		errno = 0;
+		env.freq = strtoll(arg, NULL, 10);
+		if (errno || env.freq <= 0) {
+			fprintf(stderr, "Invalid freq (in hz): %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static int nr_cpus;
+
+static int open_and_attach_perf_event(int freq, struct bpf_program *prog,
+				struct bpf_link *links[])
+{
+	struct perf_event_attr attr = {
+		.type = PERF_TYPE_SOFTWARE,
+		.freq = 1,
+		.sample_period = freq,
+		.config = PERF_COUNT_SW_CPU_CLOCK,
+	};
+	int i, fd;
+
+	for (i = 0; i < nr_cpus; i++) {
+		fd = syscall(__NR_perf_event_open, &attr, -1, i, -1, 0);
+		if (fd < 0) {
+			fprintf(stderr, "failed to init perf sampling: %s\n",
+				strerror(errno));
+			return -1;
+		}
+		links[i] = bpf_program__attach_perf_event(prog, fd);
+		if (libbpf_get_error(links[i])) {
+			fprintf(stderr, "failed to attach perf event on cpu: "
+				"%d\n", i);
+			links[i] = NULL;
+			close(fd);
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+int libbpf_print_fn(enum libbpf_print_level level,
+		    const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void sig_handler(int sig)
+{
+}
+
+static int init_freqs_hmz(__u32 *freqs_mhz, int nr_cpus)
+{
+	char *key = "cpu MHz", *p, line[32];
```

Comment:
some lines can be much longer than that, so you'd be churning through them piece by piece. How about bumping it to 128 or 256?